database-mongodb/MySQL/testDatabaseQueryEfficiency.js

```diff
@@ -0,0 +1,567 @@
+let fetchItemImages;
+let fetchMultipleItemImages;
+let connection;
+let insertRecord;
+let updateRecord;
+let deleteRecord;
+
+//Before testing async or sync mode, make sure to clear database cache and select a mode.
+//Async and sync mode are defined as:
+//Async means a call completes before the next call completes, isolating each call from each other
+//Sync means calls are made one after the other based on the requestDelay, which is in milliseconds
+//In both cases, the individual test functions (called by handleTests) are called independent
+//of each other and so no calls between tests interfere with each other
+//For both sync test modes, requestDelay also has to be uncommented
+//The last mode is 'test' mode, which should be selected when running this repo's
+//unit tests with the commmand >npm run test
+const numberOfRandomQueries = 150;
+// const mode = 'async';
+// const mode = 'sync';
+const requestDelay = 2; //requestDelay should be 2 in 'test' mode and greater than 0 for 'sync' mode
+const mode = 'test';
+
+if (mode === 'async' || mode === 'sync') {
+  const Images = require('./Images.js');
+
+  fetchItemImages = Images.fetchItemImages;
+  fetchMultipleItemImages = Images.fetchMultipleItemImages;
+  connection = Images.connection;
+  insertRecord = Images.insertRecord;
+  updateRecord = Images.updateRecord;
+  deleteRecord = Images.deleteRecord;
+}
+
+const newItemIds = [];
+const newItemData = [];
+const startingId = 10000100;
+
+const generateNewItems = function() {
+  let currentId = startingId;
+  const endId = startingId + numberOfRandomQueries - 1;
+
+  while (currentId <= endId) {
+    newItemIds.push(currentId.toString());
+    const roll = Math.ceil(Math.random() * 3);
+    const itemData = new Array(roll).fill(currentId);
+    const joinedData = itemData.join('XXX');
+    newItemData.push(joinedData);
+    currentId++;
+  }
+};
+
+const singleItemData = [];
+const generateSingleItemData = function () {
```

Comment:
A way to reduce code mirroring would be for `generateSingleItemData` to take in an array you want to mutate as well as the count. That way you could use it inside of `generateArrayItemData` instead of the inner while loop.

Reply:
Yes, also, instead of an initiation function for each test, I could prob put all the data in an array and have those arrays inside an object with the testName as the key so that one function is only needed to start each test, after it is passed the name of the test to run. Which then means I could print out the result of each test by doing something similar...but this is code that won't be used for production and so I don't think I will try to implement any refactors.

While inefficiently, it does what it is supposed to do and since it won't ever be experienced by an actual user, inefficient is OK as far as I am concerned. Do you agree?

Reply:
Gotcha, I agree with your . For the sake of efficiency of work and deadlines, whatever works is probably better than the most 'optimial' haha.